app/src/main/java/com/github/multimatum_team/multimatum/MultimatumApp.kt

```diff
@@ -1,13 +1,109 @@
 package com.github.multimatum_team.multimatum
 
+import android.app.Activity
+import android.app.ActivityManager
 import android.app.Application
+import android.content.Context
+import android.content.SharedPreferences
+import android.os.Bundle
+import com.github.multimatum_team.multimatum.LogUtil.debugLog
+import com.github.multimatum_team.multimatum.LogUtil.logFunctionCall
+import com.github.multimatum_team.multimatum.MainSettingsActivity.Companion.PROCRASTINATION_FIGHTER_ENABLED_PREF_KEY
+import com.github.multimatum_team.multimatum.service.ProcrastinationDetectorService
 import dagger.hilt.android.HiltAndroidApp
+import javax.inject.Inject
 
 @HiltAndroidApp
-class MultimatumApp: Application() {
+class MultimatumApp : Application(), Application.ActivityLifecycleCallbacks {
 
     // This class is required by Hilt
-    // And code generation seems to be very sensitive to changes in the format of this file
-    // tl;dr: BLACK MAGIC, DO NOT TOUCH
+    // It is also useful to perform actions when the app is started
+
+    override fun onCreate() {
+        logFunctionCall()
+        super.onCreate()
+        registerActivityLifecycleCallbacks(this)
+    }
+
+    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
+        logFunctionCall("created ${activity.localClassName}")
+    }
+
+    override fun onActivityStarted(activity: Activity) {
+        logFunctionCall("started ${activity.localClassName}")
+    }
+
+    override fun onActivityResumed(activity: Activity) {
+        logFunctionCall("resumed ${activity.localClassName}")
+        val onForeground = isAppOnForeground()
+        debugLog(if (onForeground) "app is on foreground" else "app is on background")
+        if (onForeground){
+            stopProcrastinationDetectorIfActive()
+        }
+    }
+
+    override fun onActivityPaused(activity: Activity) {
+        logFunctionCall("paused ${activity.localClassName}")
+    }
+
+    override fun onActivityStopped(activity: Activity) {
+        logFunctionCall("stopped ${activity.localClassName}")
+        val onForeground = isAppOnForeground()
+        debugLog(if (onForeground) "app is on foreground" else "app is on background")
+        if (!onForeground){
+            startProcrastinationDetectorIfNeeded()
+        }
+    }
+
+    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
+        logFunctionCall()
+    }
+
+    override fun onActivityDestroyed(activity: Activity) {
+        logFunctionCall("destroyed ${activity.localClassName}")
+    }
+
+    @Inject
+    lateinit var sharedPreferences: SharedPreferences
```

Comment:
Nitpick: I think it would be preferable to use the convention that all the attribute definitions (including injects) appear before the method definitions.

Reply:
Indeed, and I think that this is a bit more important than nitpicking!

Reply:
> We could probably setup the LogUtils so that the logs are only emitted on debug mode, and ignored in release mode.

I thought of it too actually :) . But how would you define debug mode? Using a constant or do you see a better solution?